File: app/src/main/java/com/druger/aboutwork/activities/LoginActivity.java
package com.druger.aboutwork.activities;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.widget.Toast;

import com.druger.aboutwork.App;
import com.druger.aboutwork.BuildConfig;
import com.druger.aboutwork.R;
import com.druger.aboutwork.db.FirebaseHelper;
import com.druger.aboutwork.model.User;
import com.druger.aboutwork.utils.Analytics;
import com.firebase.ui.auth.AuthUI;
import com.firebase.ui.auth.ErrorCodes;
import com.firebase.ui.auth.IdpResponse;
import com.google.firebase.analytics.FirebaseAnalytics;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.FirebaseUserMetadata;
import com.google.firebase.auth.UserProfileChangeRequest;

import java.util.Arrays;

import javax.inject.Inject;

import timber.log.Timber;

public class LoginActivity extends AppCompatActivity {
    private static final int RC_SIGN_IN = 1;
    private static final String TAG = LoginActivity.class.getSimpleName();

    @Inject
    Analytics analytics;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        App.Companion.getAppComponent().inject(this);
        showAuthUI();
    }

    private void showAuthUI() {
        startActivityForResult(
                AuthUI.getInstance()
                        .createSignInIntentBuilder()
                        .setAvailableProviders(Arrays.asList(
                                new AuthUI.IdpConfig.PhoneBuilder().build(),
                                new AuthUI.IdpConfig.EmailBuilder().build()
                        ))
                        .setLogo(R.drawable.ic_logo_full)
                        .setIsSmartLockEnabled(!BuildConfig.DEBUG, false)
                        .build(), RC_SIGN_IN);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == RC_SIGN_IN) {
            IdpResponse idpResponse = IdpResponse.fromResultIntent(data);

            if (resultCode == RESULT_OK) {
                analytics.logEvent(FirebaseAnalytics.Event.LOGIN);
                if (isNewUser()) {
                    saveNewUser();
                }
                startActivity(new Intent(LoginActivity.this, MainActivity.class));
                finish();
            } else {

                if (idpResponse.getError().getErrorCode() == ErrorCodes.NO_NETWORK) {
                    Toast.makeText(this, R.string.no_internet_connection, Toast.LENGTH_SHORT).show();
                    return;
                }

                Toast.makeText(this, R.string.unknown_error, Toast.LENGTH_SHORT).show();
                Timber.e(idpResponse.getError(), "Sign-in error: ");
            }
        }
    }

    private boolean isNewUser() {
        FirebaseUserMetadata metadata = FirebaseAuth.getInstance().getCurrentUser().getMetadata();
        return metadata.getCreationTimestamp() == metadata.getLastSignInTimestamp();
    }

    private void saveNewUser() {
        FirebaseUser firebaseUser = FirebaseAuth.getInstance().getCurrentUser();
        String id = firebaseUser.getUid();
        String nameEmail = firebaseUser.getDisplayName();
        String namePhone = "";
        if (nameEmail == null) {
            namePhone = "User_" + id.substring(0, 4);
            setDisplayName(firebaseUser, namePhone);
        }
        String name = nameEmail != null ? nameEmail : namePhone;
        User user = new User(id, name);
        FirebaseHelper.INSTANCE.addUser(user, id);
    }

    private void setDisplayName(FirebaseUser firebaseUser, String namePhone) {
        UserProfileChangeRequest profileUpdates = new UserProfileChangeRequest.Builder()
                .setDisplayName(namePhone).build();
        firebaseUser.updateProfile(profileUpdates);
    }


    //    @Override
//    protected void onDestroy() {
//        super.onDestroy();
//        RefWatcher refWatcher = App.Companion.getRefWatcher(this);
//        refWatcher.watch(this);
//    }
}

File: app/src/main/java/com/druger/aboutwork/activities/LoginActivity.kt
package com.druger.aboutwork.activities

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.druger.aboutwork.BuildConfig
import com.druger.aboutwork.R
import com.druger.aboutwork.db.FirebaseHelper
import com.druger.aboutwork.model.User
import com.druger.aboutwork.utils.Analytics
import com.firebase.ui.auth.AuthUI
import com.firebase.ui.auth.ErrorCodes
import com.firebase.ui.auth.IdpResponse
import com.google.firebase.analytics.FirebaseAnalytics
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.UserProfileChangeRequest
import org.koin.core.KoinComponent
import org.koin.core.inject
import timber.log.Timber

class LoginActivity : AppCompatActivity(), KoinComponent {

    private val analytics: Analytics by inject()

    private var nextScreen: String? = null
    private var companyId: String? = null
    private var reviewId: String? = null
    private var message: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        showAuthUI()
        getExtras()
    }

    private fun getExtras() {
        nextScreen = intent.getStringExtra(MainActivity.NEXT_SCREEN)
        companyId = intent.getStringExtra(MainActivity.COMPANY_ID)
        reviewId = intent.getStringExtra(MainActivity.REVIEW_ID)
        message = intent.getStringExtra(MainActivity.MESSAGE)
    }

    private fun showAuthUI() {
        startActivityForResult(
            AuthUI.getInstance()
                .createSignInIntentBuilder()
                .setAvailableProviders(listOf(
                    AuthUI.IdpConfig.PhoneBuilder().build(),
                    AuthUI.IdpConfig.GoogleBuilder().build(),
                    AuthUI.IdpConfig.AnonymousBuilder().build()
                ))
                .setLogo(R.drawable.ic_logo)
                .setIsSmartLockEnabled(!BuildConfig.DEBUG, false)
                .build(), RC_SIGN_IN)
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == RC_SIGN_IN) {
            val idpResponse = IdpResponse.fromResultIntent(data)

            if (resultCode == Activity.RESULT_OK) {
                sendAnalytics(idpResponse)
                if (isNewUser()) saveNewUser()
                startMainActivity()
                finish()
            } else {
                if (idpResponse?.error?.errorCode == ErrorCodes.NO_NETWORK) {
                    Toast.makeText(this, R.string.no_internet_connection, Toast.LENGTH_SHORT).show()
                    return
                }
                Toast.makeText(this, R.string.unknown_error, Toast.LENGTH_SHORT).show()
                Timber.e(idpResponse?.error, "Sign-in error: ")
            }
        }
    }

    private fun sendAnalytics(idpResponse: IdpResponse?) {
        analytics.logEvent(FirebaseAnalytics.Event.LOGIN)
        val type = idpResponse?.providerType
        type?.let { analytics.logEvent(Analytics.SIGN_IN_TYPE, it) }
    }

    private fun startMainActivity() {
        val intent = Intent(this@LoginActivity, MainActivity::class.java).apply {
            flags = Intent.FLAG_ACTIVITY_CLEAR_TASK or Intent.FLAG_ACTIVITY_NEW_TASK
            putExtra(MainActivity.NEXT_SCREEN, nextScreen)
            putExtra(MainActivity.COMPANY_ID, companyId)
            putExtra(MainActivity.REVIEW_ID, reviewId)
            putExtra(MainActivity.MESSAGE, message)
        }
        startActivity(intent)
    }

    private fun isNewUser(): Boolean {
        val metadata = FirebaseAuth.getInstance().currentUser?.metadata
        val diff = metadata?.lastSignInTimestamp?.minus(metadata.creationTimestamp)
        return diff == 0L || diff == 1L
    }

    private fun saveNewUser() {
        val firebaseUser = FirebaseAuth.getInstance().currentUser
        val id = firebaseUser?.uid
        val name = "User_" + id?.substring(0, 4)
        val user = id?.let { User(it, name) }
        firebaseUser?.let { setDisplayName(it, name) }
        user?.let { FirebaseHelper.addUser(it, id) }
    }

    private fun setDisplayName(firebaseUser: FirebaseUser, name: String) {
        val profileUpdates = UserProfileChangeRequest.Builder()
            .setDisplayName(name).build()
        firebaseUser.updateProfile(profileUpdates)
    }

    companion object {
        private const val RC_SIGN_IN = 1
    }


    //    @Override
    //    protected void onDestroy() {
    //        super.onDestroy();
    //        RefWatcher refWatcher = App.Companion.getRefWatcher(this);
    //        refWatcher.watch(this);
    //    }
}
